pkg/server/commands/reverseexpand/reverse_expand_weighted.go

```diff
@@ -0,0 +1,511 @@
+	mu    sync.Mutex
+}
+
+func newJobQueue() *jobQueue {
+	return &jobQueue{queue: *aq.New()}
+}
+
+func (q *jobQueue) Empty() bool {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.queue.Empty()
+}
+
+func (q *jobQueue) enqueue(value ...queryJob) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	for _, item := range value {
+		q.queue.Enqueue(item)
+	}
+}
+
+func (q *jobQueue) dequeue() (queryJob, bool) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	val, ok := q.queue.Dequeue()
+	if !ok {
+		return queryJob{}, false
+	}
+	job, ok := val.(queryJob)
+	if !ok {
+		return queryJob{}, false
+	}
+
+	return job, true
+}
+
+// loopOverEdges iterates over a set of weightedGraphEdges and acts as a dispatcher,
+// processing each edge according to its type to continue the reverse expansion process.
+//
+// While traversing, loopOverEdges appends relation entries to a stack for use in querying after traversal is complete.
+// It will continue to dispatch and traverse the graph until it reaches a DirectEdge, which
+// leads to a leaf node in the authorization graph. Once a DirectEdge is found, loopOverEdges invokes
+// queryForTuples, passing it the stack of relations it constructed on the way to that particular leaf.
+//
+// For each edge, it creates a new ReverseExpandRequest, preserving the context of the overall query
+// but updating the traversal state (the 'stack') based on the edge being processed.
+//
+// The behavior is determined by the edge type:
+//
+//   - DirectEdge: This represents a direct path to data. Here we initiate a call to
+//     `queryForTuples` to query the datastore for tuples that match the relationship path
+//     accumulated in the stack. This is the end of the traversal.
+//
+//   - ComputedEdge, RewriteEdge, and TTUEdge: These represent indirections in the authorization model.
+//     The function modifies the traversal 'stack' to reflect the next relationship that needs to be resolved.
+//     It then calls `dispatch` to continue traversing the graph with this new state until it reaches a DirectEdge.
+func (c *ReverseExpandQuery) loopOverEdges(
+	ctx context.Context,
+	req *ReverseExpandRequest,
+	edges []*weightedGraph.WeightedAuthorizationModelEdge,
+	needsCheck bool,
+	resolutionMetadata *ResolutionMetadata,
+	resultChan chan<- *ReverseExpandResult,
+) error {
+	pool := concurrency.NewPool(ctx, int(c.resolveNodeBreadthLimit))
+
+	for _, edge := range edges {
+		newReq := req.clone()
+		newReq.weightedEdge = edge
+
+		toNode := edge.GetTo()
+
+		// Going to a userset presents risk of infinite loop. Using from + to ensures
+		// we don't traverse the exact same edge more than once.
+		goingToUserset := toNode.GetNodeType() == weightedGraph.SpecificTypeAndRelation
+		if goingToUserset {
+			key := edge.GetFrom().GetUniqueLabel() + toNode.GetUniqueLabel()
+			if _, loaded := c.visitedUsersetsMap.LoadOrStore(key, struct{}{}); loaded {
+				// we've already visited this userset through this edge, exit to avoid an infinite cycle
+				continue
```

Comment:
Not a blocker but probably good to follow up on subsequent PR is to add test for this case (i am thinking of when we add in the changes to allow recursion)